Add blog post on short builtin calls (#551)
v8/v8.dev: src/blog/short-builtin-calls.md

---
title: 'Short Builtin Calls'
author: '[Toon Verwaest](https://twitter.com/tverwaes), The Big Short'
avatars:
  - toon-verwaest
date: 2021-05-06
tags:
  - JavaScript

> description: 'Faster builtin calls in V8 v9.1'

Author: done!

V8 JIT code very commonly calls into builtins. Builtins are essentially snippets of machine code that are shipped as part of the VM. There are builtins that implement full JavaScript standard library functions, such as `Function.prototype.bind`, but many builtins are helper snippets of machine code that fill in the gap between the higher-level semantics of JS and the low-level capabilities of the CPU. For example, if a JavaScript function wants to call another JavaScript function, it is common for the function implementation to call a `CallFunction` builtin that figures out how the target JavaScript function should be called; i.e., whether it’s a proxy or a regular function, how many arguments it expects, etc. Since these snippets are known when we build the VM, they are "embedded" in the Chrome binary, which means that they end up within the Chrome binary code region.

## Direct vs Indirect Calls

> On 64 bit architectures, the Chrome binary, which includes these builtins, lies arbitrarily far away from JIT code. With the [x86-64](https://en.wikipedia.org/wiki/X86-64) instruction set, this means we can’t use direct calls: they take a 32-bit signed immediate that’s used as an offset to the address of the call, and the target may be more than 2 GiB away. Instead, we need to rely on indirect calls through a register or memory operand. Such calls rely more heavily on prediction since it’s not immediately apparent from the call instruction itself what the target of the call is. On [ARM64](https://en.wikipedia.org/wiki/AArch64) we can’t use direct calls at all since the range is limited to 128 MiB. This means that in both cases we rely on the accuracy of the CPU's indirect branch predictor.

Reviewer: ```suggestion
On 64-bit architectures, the Chrome binary, which includes these builtins, lies arbitrarily far away from JIT code. With the [x86-64](https://en.wikipedia.org/wiki/X86-64) instruction set, this means we can’t use direct calls: they take a 32-bit signed immediate that’s used as an offset to the address of the call, and the target may be more than 2 GiB away. Instead, we need to rely on indirect calls through a register or memory operand. Such calls rely more heavily on prediction since it’s not immediately apparent from the call instruction itself what the target of the call is. On [ARM64](https://en.wikipedia.org/wiki/AArch64) we can’t use direct calls at all since the range is limited to 128 MiB. This means that in both cases we rely on the accuracy of the CPU's indirect branch predictor.
```

> ## Indirect Branch Prediction Limitations

Reviewer: ```suggestion
## Indirect branch prediction limitations
```

## Temporary Solution: Copy the Builtins

To avoid the cost of frequent mispredictions, and to avoid unnecessarily relying on branch prediction where possible on x86-64, we’ve decided to temporarily copy the builtins into V8's pointer compression cage on desktop machines with enough memory. This puts the copied builtin code close to dynamically generated code. The performance results heavily depend on the device configuration, but here are some results from our performance bots:

![Browsing benchmarks recorded from live pages](/_img/short-builtin-calls/v8-browsing.svg)

![Benchmark score improvement](/_img/short-builtin-calls/benchmarks.svg)

> Unembedding builtins does increase memory usage on affected devices by 1.2 to 1.4MiB per V8 instance. As a better long-term solution we’re looking into allocating JIT code closer to the Chrome binary. That way we can re-embed the builtins to regain the memory benefits, while additionally improving the performance of calls from V8-generated code to C++ code.

Reviewer: ```suggestion
Unembedding builtins does increase memory usage on affected devices by 1.2 to 1.4 MiB per V8 instance. As a better long-term solution we’re looking into allocating JIT code closer to the Chrome binary. That way we can re-embed the builtins to regain the memory benefits, while additionally improving the performance of calls from V8-generated code to C++ code.
```
